beacon-chain/state/stategen/migrate.go

```diff
@@ -14,143 +12,54 @@ import (
 // MigrateToCold advances the split point in between the cold and hot state sections.
 // It moves the recent finalized states from the hot section to the cold section and
 // only preserve the ones that's on archived point.
-func (s *State) MigrateToCold(ctx context.Context, finalizedSlot uint64, finalizedRoot [32]byte) error {
+func (s *State) MigrateToCold(ctx context.Context, fSlot uint64, fRoot [32]byte) error {
 	ctx, span := trace.StartSpan(ctx, "stateGen.MigrateToCold")
 	defer span.End()
 
-	// Verify migration is sensible. The new finalized point must increase the current split slot, and
-	// on an epoch boundary for hot state summary scheme to work.
-	currentSplitSlot := s.splitInfo.slot
-	if currentSplitSlot > finalizedSlot {
+	lastFSlot := s.splitInfo.slot
+	lastFRoot := s.splitInfo.root
+	if lastFSlot > fSlot {
 		return nil
 	}
 
-	// Migrate all state summary objects from cache to DB.
-	if err := s.beaconDB.SaveStateSummaries(ctx, s.stateSummaryCache.GetAll()); err != nil {
-		return err
-	}
-	s.stateSummaryCache.Clear()
-
-	lastArchivedIndex, err := s.beaconDB.LastArchivedIndex(ctx)
-	if err != nil {
-		return err
-	}
-
-	// Move the states between split slot to finalized slot from hot section to the cold section.
-	filter := filters.NewFilter().SetStartSlot(currentSplitSlot).SetEndSlot(finalizedSlot - 1)
-	blockRoots, err := s.beaconDB.BlockRoots(ctx, filter)
-	if err != nil {
-		return err
-	}
-
-	for _, r := range blockRoots {
-		stateSummary, err := s.beaconDB.StateSummary(ctx, r)
-		if err != nil {
-			return err
-		}
-		if stateSummary == nil || stateSummary.Slot == 0 {
-			continue
-		}
-
-		archivedPointIndex := stateSummary.Slot / s.slotsPerArchivedPoint
-		nextArchivedPointSlot := (lastArchivedIndex + 1) * s.slotsPerArchivedPoint
-		// Only migrate if current slot is equal to or greater than next archived point slot.
-		if stateSummary.Slot >= nextArchivedPointSlot {
-			// If was a skipped archival index. The node should recover previous last archived index and state.
-			if skippedArchivedPoint(archivedPointIndex, lastArchivedIndex) {
-				recoveredIndex, err := s.recoverArchivedPoint(ctx, archivedPointIndex)
-				if err != nil {
-					return err
-				}
-				lastArchivedIndex = recoveredIndex
-			}
-
-			if !s.beaconDB.HasState(ctx, r) {
-				continue
-			}
-			if err := s.beaconDB.SaveArchivedPointRoot(ctx, r, archivedPointIndex); err != nil {
+	s.finalized.lock.RLock()
```

Comment:
Instead of holding this lock for so long, can you make a copy and hold onto the finalized state?
```go
s.finalized.lock.RLock()
fs := s.finalized.state
s.finalized.lock.RUnlock()
```